docs/source/cpp/datatypes.rst

```diff
@@ -66,3 +66,138 @@ To instantiate data types, it is recommended to call the provided
    type = arrow::timestamp(arrow::TimeUnit::MICRO);
    // A list type of single-precision floating-point values
    type = arrow::list(arrow::float32());
+
+
+
+Type Traits
+-----------
+
+Writing code that can handle concrete :class:`arrow::DataType` subclasses would 
+be verbose, if it weren't for type traits. They are empty structs with type 
+declarations that map the Arrow data types to the specialized array, scalar, 
+builder, and other associated types. For example, boolean type has traits:
+
+.. code-block:: cpp
+
+   template <>
+   struct TypeTraits<BooleanType> {
+     using ArrayType = BooleanArray;
+     using BuilderType = BooleanBuilder;
+     using ScalarType = BooleanScalar;
+     using CType = bool;
+
+     static constexpr int64_t bytes_required(int64_t elements) {
+       return bit_util::BytesForBits(elements);
+     }
+     constexpr static bool is_parameter_free = true;
+     static inline std::shared_ptr<DataType> type_singleton() { return boolean(); }
+   };
+
+See the :ref:`type-traits` for an explanation of each of these fields.
+
+Using type traits, one can write template functions that can handle a variety
+of Arrow types. For example, to write a function that creates an array of 
+fibonacci values:
+
+.. code-block:: cpp
+
+   template <typename DataType,
+             typename BuilderType = typename arrow::TypeTraits<DataType>::BuilderType,
+             typename ArrayType = typename arrow::TypeTraits<DataType>::ArrayType,
+             typename CType = typename arrow::TypeTraits<DataType>::CType>
+   arrow::Result<std::shared_ptr<ArrayType>> make_fibonacci(int32_t n) {
```

Comment:
nit, but maybe try to follow the Arrow naming conventions still here (MakeFibonacci, SumArray)?